spec/mem/tree_borrows/tree.md

```diff
@@ -0,0 +1,140 @@
+    /// The `path` is the path from the `self` to the accessed node.
+    /// The `is_ancestor` indicates whether the `self` is an ancestor of the accessed node.
+    ///
+    /// This method will throw UBs, representing the undefined behavior captured by Tree Borrows.
+    fn access(
+        &mut self,
+        path: Path, // root -> child
+        access_kind: AccessKind,
+        offset_in_alloc: Size,
+        size: Size,
+        is_ancestor: bool,
+    ) -> Result {
+        // Indicates whether the accessed node is a child or foreign
+        // *from the perspective of the `self`*.
+        //
+        // If the `self` is an ancestor of the accessed node, the accessed node is a child of the `self`.
+        // Otherwise, the accessed node is a `foreign` of the `self`
+        let node_relation = if is_ancestor { NodeRelation::Child } else { NodeRelation::Foreign };
+
+        // perform state transition on the `self`
+        self.transition(node_relation, access_kind, offset_in_alloc, size)?;
+
+        let children_num = self.children.len();
+
+        // If the `self` is leaf, stop here.
+        if children_num == Int::ZERO { return ret(()); }
+
+        // The path from the sub-root
+        let sub_path = if path.len() == Int::ZERO { path } else { path.subslice_with_length(Int::ONE, path.len() - Int::ONE) };
+
+        let mut ancestor_mask = list![false; children_num];
+
+        // If the `self` is not an ancestor of the accessed node,
+        // no child of the `self` could be an ancestor of the accessed node
+        if is_ancestor {
+            if let Some(sub_root_id) = path.first() { ancestor_mask.set(sub_root_id, is_ancestor); }
+        }
+
+        self.children = self.children.zip(ancestor_mask).try_map(|(mut child, child_is_ancestor)| {
+            child.access(sub_path, access_kind, offset_in_alloc, size, child_is_ancestor)?;
+            ret::<Result<Node>>(child)
+        })?;
+
+        ret(())
+    }
+}
+```
+
+We also implement some helper methods for manipulating Trees.
+
+```rust
+impl Node {
+    /// Add a new child node to the tree whose root is self
+    /// The `path` is the path from the `self` to the parent of the `node`.
+    ///
+    /// Return the path from the `self` to the `node`
+    fn add_node(&mut self, parent_path: Path, child: Node) -> Result<Path> {
+        if parent_path.len() == Int::ZERO {
+            self.children.push(child);
+            return ret(list![self.children.len() - Int::ONE]);
+        }
+
+        // If the `self` is leaf, we fail to find the parent node.
+        if self.children.len() == Int::ZERO {
+            panic!("Tree Borrows: Not found the parent node");
+        }
+
+        // The `path` indicates which node to access next; call it the sub-root.
+        let sub_root_id = parent_path.first().unwrap();
+
+        // The path from the sub-root to the child
+        let sub_path = parent_path.subslice_with_length(Int::ONE, parent_path.len() - Int::ONE);
```

Comment:
We should add a `split_first` method to `List`, like [this one](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.split_first).